packages/predicate/src/predicate.ts

```diff
@@ -1,20 +1,41 @@
 import type { BytesLike } from '@ethersproject/bytes';
 import { arrayify } from '@ethersproject/bytes';
-import type { JsonAbiFragmentType } from '@fuel-ts/abi-coder';
+import { Logger } from '@ethersproject/logger';
+import { Interface } from '@fuel-ts/abi-coder';
+import type { JsonAbiFragmentType, JsonAbi } from '@fuel-ts/abi-coder';
 import { Address } from '@fuel-ts/address';
 import { ContractUtils } from '@fuel-ts/contract';
 import { AbstractPredicate } from '@fuel-ts/interfaces';
 import type { AbstractAddress } from '@fuel-ts/interfaces';
+import { versions } from '@fuel-ts/versions';
+
+const logger = new Logger(versions.FUELS);
 
 export class Predicate extends AbstractPredicate {
   bytes: Uint8Array;
   address: AbstractAddress;
   types?: ReadonlyArray<JsonAbiFragmentType>;
+  interface?: Interface;
 
-  constructor(bytes: BytesLike, types?: ReadonlyArray<JsonAbiFragmentType>) {
+  constructor(bytes: BytesLike, types?: JsonAbi | Interface | ReadonlyArray<JsonAbiFragmentType>) {
     super();
     this.bytes = arrayify(bytes);
     this.address = Address.fromB256(ContractUtils.getContractRoot(this.bytes));
-    this.types = types;
+
+    if (Array.isArray(types)) {
+      this.types = types;
+    } else if (types) {
+      const abiInterface = types instanceof Interface ? types : new Interface(types as JsonAbi);
+      const mainFunction = abiInterface.fragments.find(({ name }) => name === 'main');
+      if (mainFunction !== undefined) {
+        this.types = mainFunction.inputs;
+      } else {
+        logger.throwArgumentError(
+          'Cannot use ABI without "main" function',
+          'Function fragments',
+          abiInterface.fragments
+        );
+      }
+    }
```

Comment:
If we'd stop overloading the `types` parameter by accepting only `JsonAbi`:

```suggestion
    const abiInterface = new Interface(types);
    const mainFunction = abiInterface.fragments.find(({ name }) => name === 'main');
    
    if (mainFunction !== undefined) {
      this.types = mainFunction.inputs;
    } else {
      logger.throwArgumentError(
        'Cannot use ABI without "main" function',
        'Function fragments',
        abiInterface.fragments,
      );
    }
```